tools/httpbintest/src/lib.rs

```diff
@@ -0,0 +1,1304 @@
+//! [`assert()`]: struct.HttpBinTest.html#method.assert
+
+#[macro_use]
+extern crate log;
+
+use std::collections::HashMap;
+
+use quiche::h3::Header;
+
+const USER_AGENT: &str = "quiche-http3bin";
+
+// Stores the request, the expected response headers, and the actual response.
+// The assert_hdrs() method is provided for convenience to validate the received
+// headers match the expected headers.
+#[derive(Clone)]
+pub struct HttpBinReq {
+    url: url::Url,
+    hdrs: Vec<Header>,
+    body: Option<Vec<u8>>,
+    expect_resp_hdrs: Option<Vec<Header>>,
+    pub resp_hdrs: Vec<Header>,
+    resp_body: Vec<u8>,
+}
+
+impl HttpBinReq {
+    pub fn new(
+        method: &str, url: &url::Url, body: Option<Vec<u8>>,
+        expect_resp_hdrs: Option<Vec<Header>>,
+    ) -> HttpBinReq {
+        let mut path = String::from(url.path());
+        if let Some(query) = url.query() {
+            path.push('?');
+            path.push_str(query);
+        }
+
+        let mut hdrs = vec![
+            Header::new(":method", method),
+            Header::new(":scheme", url.scheme()),
+            Header::new(":authority", url.host_str().unwrap()),
+            Header::new(":path", &path),
+            Header::new("user-agent", USER_AGENT),
+        ];
+
+        if let Some(body) = &body {
+            hdrs.push(Header::new("content-length", &body.len().to_string()));
+        }
+
+        HttpBinReq {
+            url: url.clone(),
+            hdrs,
+            body,
+            expect_resp_hdrs: expect_resp_hdrs.clone(),
+            resp_hdrs: Vec::new(),
+            resp_body: Vec::new(),
+        }
+    }
+
+    pub fn assert_hdrs(&self) -> bool {
+        let mut ok = true;
+
+        if let Some(expect_hdrs) = &self.expect_resp_hdrs {
+            for hdr in expect_hdrs {
+                if let Some(h) =
+                    self.resp_hdrs.iter().find(|&x| x.name() == hdr.name())
+                {
+                    ok &= hdr.value() == h.value();
+                    continue;
+                }
+
+                ok &= false;
+            }
+        }
+
+        ok
+    }
+}
+
+// A rudimentary structure to hold httpbin response data
+#[derive(Debug, serde::Deserialize)]
+struct HttpBinResponseBody {
+    args: Option<HashMap<String, String>>,
+    data: Option<String>,
+    files: Option<HashMap<String, String>>,
+    form: Option<HashMap<String, String>>,
+    headers: Option<HashMap<String, String>>,
+    json: Option<HashMap<String, String>>,
+    #[serde(rename = "user-agent")]
+    user_agent: Option<String>,
+    server: Option<String>,
+    #[serde(rename = "content-type")]
+    content_type: Option<Vec<String>>,
+    origin: Option<String>,
+    url: Option<String>,
+}
+
+// A helper function pointer type for assertions.
+// Each test takes a vector of HttpBinReq for validation
+// and must return a pass or fail.
+type HttpBinTestAssert = (fn(&[HttpBinReq]) -> bool);
```

Comment:
```suggestion
type HttpBinTestAssert = fn(&[HttpBinReq]) -> bool;
```

I don't think you need parentheses here.